python/lsst/resources/file.py

```diff
@@ -123,7 +123,10 @@ def isdir(self) -> bool:
             `True` if this URI is a directory or looks like a directory,
             else `False`.
         """
-        return self.dirLike or os.path.isdir(self.ospath)
+        if self.dirLike is None:
+            # Cache state for next time.
+            self.dirLike = os.path.isdir(self.ospath)
```

Comment:
`ResourcePath` is nominally immutable but this cache can be updated from `None` to `bool`.  This will presumably need locking around it and in packageresource and schemeless implementations.

I am also wondering if this and isTemporary should be made read-only getters with `_dirLike` used internally.